src/SortingAlgorithms.jl

```diff
@@ -61,9 +61,15 @@ uint_mapping(o::Lt,   x     ) = error("uint_mapping does not work with general L
 const RADIX_SIZE = 11
 const RADIX_MASK = 0x7FF
 
-function sort!(vs::AbstractVector, lo::Int, hi::Int, ::RadixSortAlg, o::Ordering, ts=similar(vs))
-    # Input checking
-    if lo >= hi;  return vs;  end
+function sort!(vs::AbstractVector, lo::Int, hi::Int, ::RadixSortAlg, o::Ordering, ts=similar(vs, 0))
```

Comment:
I don't think we need `ts` as an argument here, right? I'm not aware of anywhere that passes their own `ts`, for example. I think we just move `ts = similar(vs)` down below the length check.

Reply:
I use it in https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/33, but indeed I've no idea why it's supported here given that it's not used. Anyway better leave this for another PR?

Reply:
Ah OK you mention this because of the addition of the length check. Indeed better move this inside the function then.

Reply:
The problem is that it's not clear whether this is part of the public API or not. In doubt, maybe we should keep it to avoid breaking things.

But I don't think we should silently resize `ts` disregarding its size. The cleanest solution would probably to use `nothing` as the default. Otherwise, we could allow 0-length `ts` as a special case, but that's a bit less clean (and that's one more allocation).

Reply:
Okay. That makes sense. I also misunderstood how `ts` was currently being treated. I mistakenly thought it was being silently resized when indeed it wasn't.

Reply:
I believe the now-current edb805c08f387102b34ac488f285c73077fc357c version is as strict as before the PR: throws a bounds error if `ts` or `vs` fail to contain all of `lo:hi` and otherwise messes with the contents of `ts` but not its size.

I also think the current API is identical to the old: one may pass in a correctly sized `ts` or leave that argument off.

Theoretically, the only changes this PR makes are:
1) when `hi-lo` is small radix sort is not run, and no allocations are made.
2) when `ts` needs to be allocated (almost always) an OffsetVector spanning `lo:hi` is allocated instead of a full `similar(vs)`
3) explicit low cost bounds checks at the beginning of the sorting algorithm that should cover all subsequent bounds errors (though no additional @inbounds are added)

Perhaps 2 and 3 belong in a different PR, but I think both are reasonable (assuming OffsetVectors are implemented as efficiently as Vectors)